engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java

```diff
@@ -207,4 +219,135 @@ public Atom makeModuleDoesNotExistError(String name) {
   public Atom makeNotInvokableError(Object target) {
     return notInvokableError.newInstance(target);
   }
+
+  /** Represents plain Java exception as a {@link TruffleObject}.
+   */
+  @ExportLibrary(InteropLibrary.class)
+  static final class WrapPlainException extends AbstractTruffleException {
+    private final AbstractTruffleException prototype;
+    private final Throwable original;
+
+    private WrapPlainException(Throwable cause) {
+      super(cause.getMessage(), cause, AbstractTruffleException.UNLIMITED_STACK_TRACE, null);
+      this.prototype = null;
+      this.original = cause;
+    }
+
+    private WrapPlainException(AbstractTruffleException prototype, Throwable original) {
+      super(prototype);
+      this.prototype = prototype;
+      this.original = original;
+    }
+
+    static AbstractTruffleException wrap(Throwable cause, Context ctx) {
+      var env = ctx.getEnvironment();
+      if (env.isHostException(cause)) {
+        var orig = env.asHostException(cause);
+        return new WrapPlainException((AbstractTruffleException) cause, orig);
+      } else if (cause instanceof AbstractTruffleException truffleEx) {
+        return truffleEx;
+      } else {
+        return new WrapPlainException(cause);
+      }
+    }
+
+    @ExportMessage
+    boolean hasExceptionMessage() {
+      return getMessage() != null;
+    }
+
+    @ExportMessage
+    public Object getExceptionMessage() {
+      return Text.create(getMessage());
+    }
+
+    @ExportMessage
+    String toDisplayString(boolean sideEffects) {
+      return original.toString();
+    }
+
+    @ExportMessage
+    Object getMembers(boolean includeInternal) {
+      return Array.empty();
+    }
+
+    @ExportMessage
+    boolean hasMembers() {
+      return true;
+    }
+
+    @ExportMessage
+    boolean isMemberInvocable(String member, @CachedLibrary(limit="1") InteropLibrary delegate) {
+      if (
+        "has_type".equals(member) ||
+        "getMessage".equals(member)
+      ) {
+        return true;
+      }
+      if (prototype == null) {
+        return false;
+      }
+      return delegate.isMemberInvocable(prototype, member);
+    }
+
+    @ExportMessage
+    Object invokeMember(String name, Object[] args, @CachedLibrary(limit="2") InteropLibrary iop) throws ArityException, UnknownIdentifierException, UnsupportedTypeException, UnsupportedMessageException {
+      if ("has_type".equals(name)) {
+        if (args.length != 1) {
+          throw ArityException.create(1,1,  args.length);
+        }
+        Object meta;
+        if (iop.isString(args[0])) {
+          meta = args[0];
+        } else {
+          try {
+            meta = iop.getMetaQualifiedName(args[0]);
+          } catch (UnsupportedMessageException e) {
+            meta = args[0];
+          }
+        }
+        if (!iop.isString(meta)) {
+          throw UnsupportedTypeException.create(args, "Provide class or fully qualified name of class to check");
+        }
```

Comment:
shouldn't this just be thrown in the `catch` clause?

Reply:
The goal is to propagate an `InteropException` if the argument isn't supported. Up to the caller to handle it and report it properly to the user.